File: python/stack/largest-rectangle-in-histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        max_area = 0
        st = []
        heights.append(0)
        hlen = len(heights)
        i = 0;

        while i < hlen:
            while st and heights[i] <= heights[st[-1]]:
                h = heights[st[-1]]
                st.pop()
                w = i
                if st:
                    w = i - st[-1] - 1
                if h * w > max_area:
                    max_area = h * w

            st.append(i)
            i += 1

        return max_area;

File: c/stack/largest-rectangle-in-histogram.c
int largestRectangleArea(int* heights, int heightsSize) {
    int *st;
    int top = -1;
    int maxArea = 0;
    int i;

    st = malloc(sizeof(int) * heightsSize);
    for(i = 0; i <= heightsSize; ++i){
        int curh = ((i == heightsSize) ? 0 : heights[i]);
        while(top > -1 && curh <= heights[st[top]]){
            int h = heights[st[top--]];
            int w = ((top == -1) ? i : (i - st[top] - 1));

            if(h * w > maxArea)
                maxArea = h * w;
        }

        st[++top] = i;
    }

    free(st);
    return maxArea;
}
